File: scripts/bindings/ggplot_distribution_geometries.py
"""FIX-GG09 independent Python precomputed distribution geometry authors."""
import json,sys,math
from pathlib import Path
ROOT=Path(__file__).resolve().parents[2]
sys.path[:0]=[str(ROOT/'packages/python'),str(Path(sys.argv[1]).resolve())]
import finstack_chart as c
out=Path(sys.argv[2]);out.mkdir(parents=True,exist_ok=True)
output=c.Output((ROOT/'fixtures/capability/fonts/NotoSans-Regular.ttf').read_bytes())
red={'red':255,'green':0,'blue':0,'alpha':255};gold={'red':255,'green':215,'blue':0,'alpha':255}
for mode in range(15):
    if mode>=11:
        data=c.Data.columns({'x':[0.,0.,.2,1.,1.2,2.,2.,2.]});mapping=c.aes()
        layer=c.density().stat(c.density_stat().input('x')).recipe({'Density':{'outline':['Upper','Lower','Both','Full'][mode-11]}}).fill(gold).alpha(.2)
    elif mode>=8:
        data=c.Data.columns({'x':[1.]*8,'y':[0.,1.,1.,2.,3.,4.,5.,20.]})
        mapping=c.aes()
        layer=c.boxplot().stat(c.boxplot_stat().input(data.field('y')).x('x')) if mode==8 else c.violin().stat(c.violin_stat().input(c.source_expr(data.field('y'))).x('x')) if mode==9 else c.dotplot().stat(c.dotplot_stat().input('y'))
    elif mode<4:
        data=c.Data.columns({'x':[1.,2.],'middle':[2.5,3.],'lo':[.75,2.],'hi':[3.25,4.],'min':[0.,1.],'max':[4.,5.],'nl':[1.1034641071565687,1.5868050382201329],'nu':[3.8965358928434313,4.413194961779867],'relative':[math.sqrt(8),math.sqrt(5)]},keys=[10,20])
        spec={'width':.75,'notch':mode==1,'notch_width':.3,'staple_width':0. if mode==0 else .7,'variable_width':mode==2,'source_outliers':[{'row':'10','values':[20.]}]}
        if mode==2:spec.update(median={'color':red,'linewidth':1.},whisker={'line_type':'Dashed'},outlier={'shape':21,'fill':gold,'size':3.})
        layer=c.rule().recipe({'Boxplot':spec})
        for channel,field in [('Lower','lo'),('Upper','hi'),('Middle','middle'),('WhiskerLower','min'),('WhiskerUpper','max'),('NotchLower','nl'),('NotchUpper','nu'),('RelativeWidth','relative')]:layer=layer.recipe_value(channel,field)
        mapping=c.aes().x('x').y('middle').group('x')
        if mode==3:mapping=c.aes().x('middle').y('x').group('x');layer=layer.orientation('Horizontal')
    elif mode==4:
        data=c.Data.columns({'x':[1.]*5,'y':[0.,1.,2.,3.,4.],'width':[.1,.7,1.,.5,.1],'q':[None,.25,.5,.75,None]},keys=[1,2,3,4,5])
        mapping=c.aes().x('x').y('y');layer=c.rule().recipe({'Violin':{'width':.8,'quantile':{'line_type':'Dashed','color':red}}}).recipe_value('ViolinWidth','width').recipe_value('QuantileFlag','q')
    else:
        data=c.Data.columns({'bin':[0.,1.,2.],'count':[3.,1.,4.]},keys=[1,2,3]);mapping=c.aes().x(1.).y('bin') if mode==6 else c.aes().x('bin').y(0.)
        layer=c.rule().recipe({'Dotplot':{'bin_axis':'Y' if mode==6 else 'X','stack':'CenterWhole' if mode==7 else 'Up','stack_ratio':.8,'dot_size':.7}}).recipe_value('Count','count').recipe_value('BinWidth',.5)
    p=c.plot(data).profile('Ggplot2_4_0_3').aes(mapping).layer(layer).build();wire=p.to_json();restored=c.Plot.from_json(wire)
    request=output.request(restored,c.export_options(600.,360.).dpi(144));frame=request.prepare();scene=frame.scene()
    original_request=output.request(p,c.export_options(600.,360.).dpi(144));original_frame=original_request.prepare();assert original_frame.scene()==scene
    original_frame.dispose();original_request.dispose()
    (out/f'distribution-{mode}.plot.json').write_text(wire);(out/f'distribution-{mode}.scene.json').write_text(json.dumps(scene))
    for fmt in ['svg','pdf','png']:(out/f'distribution-{mode}.{fmt}').write_bytes(frame.export(fmt))
    frame.dispose();request.dispose();restored.dispose();p.dispose();data.dispose()
output.dispose();print('PASS Python distribution geometry: fifteen authors, 45 publications.')

File: examples/common/ggplot_distribution_geometries.rs
//! GG09 precomputed source geometry and control contracts; estimator authors are separate.
use chart_core::plot::{boxplot, boxplot_stat, dotplot, dotplot_stat, violin, violin_stat};
use chart_core::{RowKey, grammar::*, prelude::*, scene::Color};
pub fn author(mode: usize) -> chart_core::ChartResult<Plot> {
    if mode >= 11 {
        let data = Data::columns()
            .column("x", [0., 0., 0.2, 1., 1.2, 2., 2., 2.])
            .build()?;
        let outline = [
            AreaOutline::Upper,
            AreaOutline::Lower,
            AreaOutline::Both,
            AreaOutline::Full,
        ][mode - 11];
        return plot(data)
            .profile(Profile::Ggplot2_4_0_3)
            .layer(
                chart_core::plot::density()
                    .stat(chart_core::plot::density_stat().input("x"))
                    .recipe(BuiltinRecipe::Density(DensityRecipe { outline }))
                    .fill(Color {
                        red: 255,
                        green: 215,
                        blue: 0,
                        alpha: 255,
                    })
                    .alpha(0.2),
            )
            .build();
    }
    if mode >= 8 {
        let data = Data::columns()
            .column("x", [1.; 8])
            .column("y", [0., 1., 1., 2., 3., 4., 5., 20.])
            .build()?;
        let layer = match mode {
            8 => boxplot().stat(boxplot_stat().input(data.field("y")?).x("x")),
            9 => violin().stat(violin_stat().input(source_expr(data.field("y")?)).x("x")),
            _ => dotplot().stat(dotplot_stat().input("y")),
        };
        return plot(data)
            .profile(Profile::Ggplot2_4_0_3)
            .layer(layer)
            .build();
    }
    if mode < 4 {
        let data = Data::columns()
            .keys([10, 20])
            .column("x", [1., 2.])
            .column("middle", [2.5, 3.])
            .column("lo", [0.75, 2.])
            .column("hi", [3.25, 4.])
            .column("min", [0., 1.])
            .column("max", [4., 5.])
            .column("nl", [1.1034641071565687, 1.5868050382201329])
            .column("nu", [3.8965358928434313, 4.413194961779867])
            .column("relative", [8_f64.sqrt(), 5_f64.sqrt()])
            .build()?;
        let mut spec = BoxplotRecipe {
            width: Some(0.75),
            notch: mode == 1,
            notch_width: 0.3,
            staple_width: if mode == 0 { 0. } else { 0.7 },
            variable_width: mode == 2,
            source_outliers: vec![SourceBoxOutliers {
                row: RowKey::new(10),
                values: vec![20.],
            }],
            ..Default::default()
        };
        if mode == 2 {
            spec.median.color = Some(
                Color {
                    red: 255,
                    green: 0,
                    blue: 0,
                    alpha: 255,
                }
                .into(),
            );
            spec.median.linewidth = Some(1.);
            spec.whisker.line_type = Some(LineType::Dashed);
            spec.outlier.shape = Some(21);
            spec.outlier.fill = Some(
                Color {
                    red: 255,
                    green: 215,
                    blue: 0,
                    alpha: 255,
                }
                .into(),
            );
            spec.outlier.size = Some(3.);
        }
        let mut layer = rule()
            .recipe(BuiltinRecipe::Boxplot(Box::new(spec)))
            .recipe_value(RecipeAesthetic::Lower, "lo")
            .recipe_value(RecipeAesthetic::Upper, "hi")
            .recipe_value(RecipeAesthetic::Middle, "middle")
            .recipe_value(RecipeAesthetic::WhiskerLower, "min")
            .recipe_value(RecipeAesthetic::WhiskerUpper, "max")
            .recipe_value(RecipeAesthetic::NotchLower, "nl")
            .recipe_value(RecipeAesthetic::NotchUpper, "nu")
            .recipe_value(RecipeAesthetic::RelativeWidth, "relative");
        let mut aes = aes().x("x").y("middle").group("x");
        if mode == 3 {
            aes = aes.x("middle").y("x");
            layer = layer.orientation(Orientation::Horizontal);
        }
        return plot(data)
            .profile(Profile::Ggplot2_4_0_3)
            .aes(aes)
            .layer(layer)
            .build();
    }
    if mode == 4 {
        let data = Data::columns()
            .keys([1, 2, 3, 4, 5])
            .column("x", [1.; 5])
            .column("y", [0., 1., 2., 3., 4.])
            .column("width", [0.1, 0.7, 1., 0.5, 0.1])
            .column("q", [None, Some(0.25), Some(0.5), Some(0.75), None])
            .build()?;
        return plot(data)
            .profile(Profile::Ggplot2_4_0_3)
            .aes(aes().x("x").y("y"))
            .layer(
                rule()
                    .recipe(BuiltinRecipe::Violin(ViolinRecipe {
                        width: Some(0.8),
                        quantile: IntervalStroke {
                            line_type: Some(LineType::Dashed),
                            color: Some(
                                Color {
                                    red: 255,
                                    green: 0,
                                    blue: 0,
                                    alpha: 255,
                                }
                                .into(),
                            ),
                            ..Default::default()
                        },
                    }))
                    .recipe_value(RecipeAesthetic::ViolinWidth, "width")
                    .recipe_value(RecipeAesthetic::QuantileFlag, "q"),
            )
            .build();
    }
    let data = Data::columns()
        .keys([1, 2, 3])
        .column("bin", [0., 1., 2.])
        .column("count", [3., 1., 4.])
        .build()?;
    let axis = if mode == 6 { DotAxis::Y } else { DotAxis::X };
    let stack = if mode == 7 {
        DotStack::CenterWhole
    } else {
        DotStack::Up
    };
    plot(data)
        .profile(Profile::Ggplot2_4_0_3)
        .aes(if axis == DotAxis::X {
            aes().x("bin").y(0.)
        } else {
            aes().x(1.).y("bin")
        })
        .layer(
            rule()
                .recipe(BuiltinRecipe::Dotplot(DotplotRecipe {
                    bin_axis: axis,
                    stack,
                    stack_ratio: 0.8,
                    dot_size: 0.7,
                    ..Default::default()
                }))
                .recipe_value(RecipeAesthetic::Count, "count")
                .recipe_value(RecipeAesthetic::BinWidth, 0.5),
        )
        .build()
}
